armsr: add 6.12 as a testing kernel version

mcbridematt opened the pull request:
This is a fairly straightforward update to 6.12 for armsr.

I have tested both armv7 and armv8 in QEMU and other platforms shown below.

~Note that there is a pending [pull request](https://github.com/openwrt/openwrt/pull/17012) to fix the serial console on device tree systems. The test status for all DT systems in the list below assumes this fix is applied.~ Serial console fix applied, this PR refreshes it for 6.12.

The reference U-Boot (package/boot/uboot-armsr) is also due for an update, a PR for that ~will be posted shortly~ is now here: https://github.com/openwrt/openwrt/pull/18862

Machine | Status
-- | --
QEMU armv7 U-Boot | OK
QEMU armv8 U-Boot | OK
QEMU armv8 EDKII/ACPI | OK
AWS a1.metal | [OK](https://gist.github.com/mcbridematt/12b4d961235b5217e8388e0d05fd0f08#file-a1-metal-txt)*
AWS t4g.nano | [OK](https://gist.github.com/mcbridematt/12b4d961235b5217e8388e0d05fd0f08#file-t4g-nano-txt)
AWS c6g.metal | [OK](https://gist.github.com/mcbridematt/12b4d961235b5217e8388e0d05fd0f08#file-c6g-metal-txt)
Ten64 (LS1088A) | OK
Raspberry Pi 4 (U-Boot rpi_64_defconfig) | Pending
Raspberry Pi 4 (PiTF/EDKII/ACPI) | Pending
Coral Dev Board (imx8mp) | Pending
Hyper-V (WindowsOnArm/SDX) | OK

<small>* a1.metal has some stack traces in dmesg and errors related to PCIe windows, but it does boot and is accessible over SSH. The new generation instances (both VM and Metal) look clean</small>

mcbridematt committed:
armsr: enable 6.12 as a testing kernel version

Allow the armsr target to be built with a 6.12 kernel
when CONFIG_TESTING_KERNEL is set.

Signed-off-by: Mathew McBride <matt@traverse.com.au>
Link: https://github.com/openwrt/openwrt/pull/18849
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

mcbridematt committed:
kernel/armsr: Restore kernel files for v6.6

This is an automatically generated commit which aids following Kernel patch
history, as git will see the move and copy as a rename thus defeating the
purpose.

For the original discussion see:
https://lists.openwrt.org/pipermail/openwrt-devel/2023-October/041673.html

Signed-off-by: Mathew McBride <matt@traverse.com.au>
Link: https://github.com/openwrt/openwrt/pull/18849
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

mcbridematt committed:
kernel/armsr: Create kernel files for v6.12 (from v6.6)

This is an automatically generated commit.

When doing `git bisect`, consider `git bisect --skip`.

Signed-off-by: Mathew McBride <matt@traverse.com.au>
Link: https://github.com/openwrt/openwrt/pull/18849
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

mcbridematt committed:
armsr: patches: refresh printk console patch for 6.12

This carries the text console behaviour fix as described in
[1] over to the 6.12 kernel.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

[1] https://github.com/openwrt/openwrt/pull/17012#issuecomment-2591751115
Link: https://github.com/openwrt/openwrt/pull/18849

Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

hauke commented on target/linux/armsr/armv7/config-6.12:
```diff
@@ -14,12 +14,14 @@ CONFIG_ARCH_SELECT_MEMORY_MODEL=y
 CONFIG_ARCH_VIRT=y
 CONFIG_ARM=y
 CONFIG_ARM_CPU_SUSPEND=y
+# CONFIG_ARM_DEBUG_WX is not set
```
This is not needed any more, it was added in aaf89c2c2b2ca8575b367816bc5081eca403a1ca

guidosarducci commented:
Very strange. I was building for armv7 rather than armv8, but  instead of the shellcheck error I see ones related to libbpf as below. Is it the same building armv7 for you?
```
make -C /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/lib/bpf OUTPUT=/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/ \
        DESTDIR=/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf prefix= \
        ARCH= CROSS_COMPILE= CC="gcc" LD="ld" AR="ar" /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/libbpf.a install_headers
gcc: error: unrecognized command-line option '-fhonour-copts'; did you mean '-fhonor-std'?
gcc: error: unrecognized command-line option '-mfloat-abi=hard'
/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/scripts/bpf_doc.py --header \
        --file /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/include/uapi/linux/bpf.h > /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/bpf_helper_defs.h
mkdir -p /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/staticobjs/
make -C /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/build CFLAGS= LDFLAGS= OUTPUT=/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/staticobjs/ /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/staticobjs/fixdep
gcc -O2 -I/home/kodidev/openwrt-project/staging_dir/host/include -I/home/kodidev/openwrt-project/staging_dir/hostpkg/include -I/home/kodidev/openwrt-project/staging_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/host/include -I/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/include  -o /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/staticobjs/fixdep /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/build/fixdep.c
make -f /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/build/Makefile.build dir=. obj=libbpf OUTPUT=/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/staticobjs/
  gcc -Wp,-MD,/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/staticobjs/.libbpf.o.d -Wp,-MT,/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/staticobjs/libbpf.o -Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -mfloat-abi=hard -fmacro-prefix-map=/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi=perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -I/home/kodidev/openwrt-project/staging_dir/toolchain-arm_cortex-a15+neon-vfpv4_gcc-14.3.0_musl_eabi/usr/include -I/home/kodidev/openwrt-project/staging_dir/toolchain-arm_cortex-a15+neon-vfpv4_gcc-14.3.0_musl_eabi/include -I/home/kodidev/openwrt-project/staging_dir/toolchain-arm_cortex-a15+neon-vfpv4_gcc-14.3.0_musl_eabi/include/fortify -std=gnu89 -Wbad-function-cast -Wdeclaration-after-statement -Wformat-security -Wformat-y2k -Winit-self -Wmissing-declarations -Wmissing-prototypes -Wnested-externs -Wno-system-headers -Wold-style-definition -Wpacked -Wredundant-decls -Wstrict-prototypes -Wswitch-default -Wswitch-enum -Wundef -Wwrite-strings -Wformat -Wno-type-limits -Wstrict-aliasing=3 -Wshadow -Wno-switch-enum -Werror -Wall -I/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/ -I/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/include -I/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/include/uapi -I/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/arch/x86/include -fvisibility=hidden -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -D"BUILD_STR(s)=#s" -c -o /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/staticobjs/libbpf.o libbpf.c
gcc: error: unrecognized command-line option '-fhonour-copts'; did you mean '-fhonor-std'?
gcc: error: unrecognized command-line option '-mfloat-abi=hard'
make[7]: *** [/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/build/Makefile.build:106: /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/staticobjs/libbpf.o] Error 1
make[6]: *** [Makefile:165: /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/staticobjs/libbpf-in.o] Error 2
make[5]: *** [Makefile:54: /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/libbpf.a] Error 2
make[4]: *** [Makefile.perf:1209: /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/bpftool] Error 2
make[3]: *** [Makefile.perf:292: sub-make] Error 2
make[2]: *** [Makefile:93: /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/.built] Error 2
make[2]: Leaving directory '/home/kodidev/openwrt-project/package/devel/perf'
time: package/devel/perf/compile#9.61#2.90#13.05
    ERROR: package/devel/perf failed to build.
make[1]: *** [package/Makefile:189: package/devel/perf/compile] Error 1
make[1]: Leaving directory '/home/kodidev/openwrt-project'
make: *** [/home/kodidev/openwrt-project/include/toplevel.mk:233: package/perf/compile] Error 2

```

mcbridematt committed:
config: remove kernel version restriction for ARM_PMUV3

The kernel CONFIG_ARM_PMUV3 option was not being presented
for 6.12 kernels with armsr/armv7.

As the minimum kernel version is now 6.6, there is no need
to limit this to specific kernel versions anymore.

This caused a kernel compile stoppage when attempting to
compile the 'perf' tool for armsr/armv7.

Signed-off-by: Mathew McBride <matt@traverse.com.au>
Link: https://github.com/openwrt/openwrt/pull/18849
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

mcbridematt committed:
armsr: 6.12: add new KConfigs and refresh

This includes both new additions to 6.12
and an attempted "refresh" of the config
to remove duplicates from target/generic/config-6.12.

(The OpenWrt kernel_makeoldconfig does not
work well with the armv8 subtarget for reasons
I am yet to determine, so that file has been
pruned manually)

Most new options are in armv8, where the
KConfig relates to a platform that will likely
be armsr compatible (like the i.MX91/93/95),
it has been enabled.

Signed-off-by: Mathew McBride <matt@traverse.com.au>
Link: https://github.com/openwrt/openwrt/pull/18849
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

guidosarducci commented:
Thanks, and apologies -- I totally forgot about that PMU fix amongst a giant pile-o'-patches...

mcbridematt commented:
> Is it the same building armv7 for you?

I can reproduce this issue when `clang` is installed on the host.

It looks like it doesn't try to use the clang that is compiled as part of OpenWrt's llvm-bpf (./host/llvm-bpf-18.1.7.Linux-x86_64/bin/clang) and that it might be mixing host vs target CFLAGS?

douglarek commented:
> > > Is it the same building armv7 for you?
> > 
> > 
> > I can reproduce this issue when `clang` is installed on the host.
> 
> OK, interesting. By default I configure OpenWrt to use the host toolchain for BPF so didn't notice, but it suggests a connection to BPF CO-RE feature testing/build.
> 
> > It looks like it doesn't try to use the clang that is compiled as part of OpenWrt's llvm-bpf (./host/llvm-bpf-18.1.7.Linux-x86_64/bin/clang) and that it might be mixing host vs target CFLAGS?
> 
> Does also seem like a compiler flags issue, since gcc below doesn't recognize ARM floating-point flags...
> 
> ```
>   CC      /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armsr_armv7/linux-6.12.31/tools/perf-target-arm_cortex-a15+neon-vfpv4_musl_eabi/util/bpf_skel/.tmp/bootstrap/libbpf/staticobjs/libbpf.o
> gcc: error: unrecognized command-line option '-fhonour-copts'; did you mean '-fhonor-std'?
> gcc: error: unrecognized command-line option '-mfloat-abi=hard'
> ```
> 
> and I also see the same thing building `malta/be32`, so it appears target CFLAGS may be passed to the host gcc...
> 
> ```
>   CC      /home/kodidev/openwrt-project/build_dir/target-mips_24kc_musl/linux-malta_be/linux-6.12.31/tools/perf-target-ips_24kc_musl/util/bpf_skel/.tmp/bootstrap/libbpf/staticobjs/libbpf.o
>         CC rehash.c
> gcc: error: unrecognized command-line option '-mno-branch-likely'
> gcc: error: unrecognized command-line option '-mips32r2'
> gcc: error: unrecognized command-line option '-fhonour-copts'; did you mean '-fh
> onor-std'?
> ```

I also encountered this issue when building the snapshot version of r5c.

[logs_41227408789.zip](https://github.com/user-attachments/files/21084047/logs_41227408789.zip)
